### PR TITLE
Some 64bit improvements, lots of 64bit TODOs/STUBBED()

### PR DESCRIPTION
**Pull request by DanielGibson**
This should fix some more 64bit issues.

I also added some more FIXMEs and STUBBED("64-but issue") where the compiler gave me corresponding warnings.

I was able to play the complete first level of the second encounter, however I think there's still a lot missing.

Does anyone know what `CShell::GetINDEX()` is about? Its return value often is casted to pointer and dereferenced directly afterwards (e.g. in Engine/Network/Network.cpp `CacheShadows()`; I didn't mark all occurences in the code).
Is there some kind of trick (like in `ModelPolygonVertex::Read_t()` where the "pointers" are actually indexes that are later converted back to pointers) that I'm missing? I'm surprised that INDEX is used to store a pointer there, not ULONG like in several other places..

### CODE REVIEW

**Comment by yamgent** [comment]
I am not officially with Croteam so you need to take this with a grain of salt.

I don't think that it is some kind of indices trick that you mentioned, but rather it is just as what you read in the code: it is merely storing a pointer to the CWorld object.

While the CShell class is typically used for storing console shell variables (in which _only some_ of these variables can be accessed by the user to do stuff such as cheating), it also serves as a way for the game to pass data around different classes, so it conveniently acts like a list of global variables.

Since INDEX and ULONG are of "similar" types (i.e. uses 32-bits in Windows 32-bits), I don't think they would bother with implementing another function just to deal with ULONG, and just cheat by using the  INDEX version instead. Of course, that's going to cause problems with your 64-bit goal here ;).

**Comment by icculus** [comment]
Yeah, INDEX is used for several things in CShell (it's an integer for numeric cvars, all the boolean values are INDEX, too (0 or 1), and occasionally they use it to store pointers cast to integer).

I suspect some pointers ended up in there because it made it easy to pass arbitrary data to shared libraries? Just a guess.

File: Sources/EntitiesMP/Common/Particles.cpp [hunk 1]
```diff
@@ -1591,7 +1591,7 @@ INDEX Particles_Regeneration(CEntity *pen, FLOAT tmStart, FLOAT tmStop, FLOAT fY
     
     UBYTE ubR = (UBYTE) (192+afStarsPositions[iRnd][1]*64);
     UBYTE ubG = (UBYTE) (192+afStarsPositions[iRnd][2]*64);
-    UBYTE ubB = (UBYTE) (192+afStarsPositions[iRnd][3]*64);
+    UBYTE ubB = (UBYTE) (192+afStarsPositions[iRnd][3]*64); // FIXME DG: afStarsPositions[i][3] invalid, what's going on here?
```
Comment (icculus) [inline Sources/EntitiesMP/Common/Particles.cpp]: I saw Clang fire this warning, too...I think this is a bug in the original code, and these should have counted from 0 to 2, not 1 to 3, but that's just a guess.

File: Sources/SeriousSam/Menu.cpp [hunk 2]
```diff
@@ -1187,7 +1187,7 @@ void JoinNetworkGame(void)
       _pNetwork->ga_strRequiredMod.ScanF("%250[^\\]\\%s", &strModName, &strModURL);
       _fnmModSelected = CTString(strModName);
       _strModURLSelected = strModURL;
-      if (_strModURLSelected="") {
+      if (_strModURLSelected="") { // FIXME DG: not == ?
```
Comment (icculus) [inline Sources/SeriousSam/Menu.cpp]: This was probably supposed to be ==.

File: Sources/CMakeLists.txt [hunk 3]
```diff
@@ -110,6 +113,7 @@ if(CMAKE_CXX_COMPILER_ID MATCHES "Clang")
     add_definitions(-Wno-switch)
     add_definitions(-Wno-tautological-undefined-compare)
     add_definitions(-Wno-c++11-compat-deprecated-writable-strings)
+    add_definitions(-Wno-logical-op-parentheses)
```
Comment (icculus) [inline Sources/CMakeLists.txt:116]: These are actually legit bugs being reported, though, and I'd like them to be cleaned up at some point.

**Comment by icculus** [comment]
Also, getting a CRC of a pointer is asking for trouble; where does it end up needing that?

File: Sources/GameMP/Controls.cpp [hunk 4]
```diff
@@ -227,8 +227,9 @@ void CControls::Load_t( CTFileName fnFile)
       achrActionName[ 0] = 0;
       FLOAT fSensitivity = 50;
       FLOAT fDeadZone = 0;
+      // FIXME DG: gcc and clang say the followign has too many arguments for the given format
       sscanf( achrLine, "%*[^\"]\"%1024[^\"]\"%*[^\"]\"%1024[^\"]\" %g %g %1024s %1024s",
-              achrActionName, achrAxis, &fSensitivity, &fDeadZone, achrIfInverted, achrIfRelative, achrIfSmooth);
+              achrActionName, achrAxis, &fSensitivity, &fDeadZone,   achrIfInverted, achrIfRelative, achrIfSmooth);
```
Comment (DrItanium) [inline Sources/GameMP/Controls.cpp]: I took a look at this as clang's warning perplexed me so I extracted this code out to see the warning. It looks like the sscanf is missing a %1024s at the end of the format string.

File: Sources/CMakeLists.txt (hunk 3 above)
Comment (DanielGibson) [inline Sources/CMakeLists.txt:116]: the problem with those are that the code is /probably/ correct as it is, and if it's not then adding incorrect braces makes the issue even harder to find..  
still may be worthwhile to fix those warnings once everything seems stable :)

File: Sources/EntitiesMP/Common/Particles.cpp (hunk 1 above)
Comment (DanielGibson) [inline Sources/EntitiesMP/Common/Particles.cpp]: probably.. but there is a remote possibility that they thought this was a good trick to get `afStarsPositions[iRnd+1][0]`? (it's not really, as for iRnd == 1023 that'd read out of bounds)

**Comment by DanielGibson** [comment]
The CRC of a pointer is used in Engine/Graphics/DrawPort.cpp `CDrawPort::GetID()`

**Comment by DanielGibson** [comment]
So, what should we do about `CShell::GetINDEX()`? introduce a `CShell::GetIntPtr()` (and all the other code to make that work)?  
or just make `GetINDEX()` return (and internally use etc) `size_t` (or probably `ssize_t`)?

**Comment by DanielGibson** [comment]
ok, I /think/ GetINDEX() is only used as a pointer for "pwoCurrentWorld", so the cases should be easy to find, and maybe we could even special case them with something like `void CShell::SetCurrentWorld(CWorld *pwo)` and `CWorld* CShell:GetCurrentWorld(void)` instead of adding an additional `ShellTypeType` etc

I haven't even tried to understand the serious sam cvar system yet, could the easy way with `GetCurrentWorld()` work or would everything break apart if currentworld is not in a named console variable?

File: Sources/EntitiesMP/Common/Particles.cpp (hunk 1 above)
Comment (icculus) [inline Sources/EntitiesMP/Common/Particles.cpp]: I guess we could just try correcting it to 0-2 and see if anything explodes.  :)

Comment (DanielGibson) [inline Sources/EntitiesMP/Common/Particles.cpp]: any idea what kind of explosion we should look for? something with colors of particles?

File: Sources/CMakeLists.txt (hunk 3 above)
Comment (icculus) [inline Sources/CMakeLists.txt:116]: Agreed. Some of them are almost certainly wrong--albeit in harmless ways, probably--but a lot of them are difficult to decide the original programmer's intent...or the game relies on them being wrong, etc.

At the minimum, we should just put in the parentheses that make Clang happy eventually, but let's do that in another PR some other time. For now, I'll settle for a MESSAGE(WARNING "reenable this warning some day") line in CMakeLists.txt ... so we don't forget to revisit, but the warning spam during the build goes away for now.

File: Sources/GameMP/Controls.cpp (hunk 4 above)
Comment (icculus) [inline Sources/GameMP/Controls.cpp]: I couldn't parse that string myself, so I was hoping someone else would for me.  :)

**Comment by icculus** [comment]
I wouldn't add a pointer type to CShell...I think mostly these are meant to be persistent vars that live in config files (and maybe get sent over the network?), so there's really no place it's safe to be passing pointers around in there. If it's really just one place doing it, let's just have a non-CShell solution (even just a global variable, if that makes sense to do. CShell::SetCurrentWorld() might even be overkill if this doesn't otherwise have anything to do with CShell. I'll leave it up to you, though.)

As for the CRC: I guess they're just hashing the pointer down to a unique value, not actually using it for a data checksum, so that's fine with me. I guess that's useful anyhow, since a CRC will be 32-bits even if the pointer is 64.

File: Sources/GameMP/Controls.cpp (hunk 4 above)
Comment (DanielGibson) [inline Sources/GameMP/Controls.cpp]: me too =)

adding `%1024s` would change the behavior (sscanf would expect one more string than it does now), so I'll comment out achrIfSmooth (which seems unused below) and add a comment

File: Sources/EntitiesMP/Common/Particles.cpp (hunk 1 above)
Comment (DanielGibson) [inline Sources/EntitiesMP/Common/Particles.cpp]: ok, now I'm really confused, found that code in the same file:

``` c++
    FLOAT fRndAppearX = afStarsPositions[iRnd][0]*vStretch(1);
    FLOAT fRndSpeedY = (afStarsPositions[iRnd][1]+0.5f)*0.125f*vStretch(2);
    FLOAT fRndAppearZ = afStarsPositions[iRnd][2]*vStretch(3);
    FLOAT3D vRndDir=FLOAT3D(afStarsPositions[iRnd][1],0,afStarsPositions[iRnd][3]);
```

there indices 0-3 are all used within 4 lines

OTOH, all those values seem to be completely random (see `InitParticleTables()`) so it probably doesn't matter

**Comment by DanielGibson** [comment]
pushed some more fixes for most (all?) of the things we discussed here, but there is still more to do, like several more tag-things from BSP.  
(probably not today, though).

Thanks @rcgordon, @DrItanium and @yamgent for the feedback! :-)

**Comment by DanielGibson** [comment]
ok, one more fix for fishy things in Texture.cpp :-)

**Comment by icculus** [comment]
Merging, thanks!

**Comment by rohit-n** [comment]
After the merge, the game now segfaults immediately when starting at line 1442 of Texture.cpp. Apparently, QtCreator considers td_pulObjects to have "no such value". (?)

**Comment by DanielGibson** [comment]
@rohit-n weird.. I have improved (I hope) the stuff in Textures.cpp a bit, can you try out the code from #29 ?

**Comment by rohit-n** [comment]
As I mentioned in #29, the segfault in Texture.cpp appears to resolved with that change. However, I am now unable to compile on my 32-bit Ubuntu 12.04 machine (x86 ASM disabled) with GCC 4.6.3, due to the changes in Types.h. Specifically, UINTPTR_MAX is not defined. Apparently in my stdint.h, the check for 

`#if !defined __cplusplus || defined __STDC_LIMIT_MACROS`

fails here, so it is never defined.

**Comment by DanielGibson** [comment]
grrr, why is the world so broken?! :-/

can you try adding `#define __STDC_LIMIT_MACROS` before `#include <stdint.h>` in Sources/Engine/Base/Types.h aroudn line 73?

**Comment by Yamagi** [comment]
From my FreeBSD adventures yesterday I can tell you that defining `__STDC_LIMIT_MACROS` right before stdint.h in Types.h is not enough. That include statement is not always the first to include stdint.h. Adding it to the top of Engine/StdH.h help in some cases, but not for sources generated by ecc. 

I choose the easy route and forced a --std=c++11 into CXXFLAGS. At least with FreeBSDs base system Clang 3.4 `UINTPTR_MAX` is available / imported when the compiler runs in C++ 11 mode.

**Comment by DanielGibson** [comment]
not sure if -std=c++11 helps on gcc 4.6 on an old ubuntu from 2012, but it might be worth a try

**Comment by rohit-n** [comment]
std=c++11 definitely does not work here.

Defining `__STDC_LIMIT_MACROS` in Types.h gets me linking the 3 .so files, but fails when compiling BaseEvents.es.

```
In file included from /home/rohit/Code/Serious-Engine/Sources/./Engine/StdH.h:51:0,
                 from /home/rohit/Code/Serious-Engine/Sources/Engine/Classes/BaseEvents.es:18:
/home/rohit/Code/Serious-Engine/Sources/./Engine/Base/Types.h:84:6: error: #error Your system does not provide UINTPRT_MAX, find another way!
```

I tried defining it again in StdH.h before the Types.h include, but the error persists.

**Comment by DanielGibson** [comment]
in the CMakeLists.txt try adding `add_definitions(-D__STDC_LIMIT_MACROS=1)` somewhere, maybe below `add_definitions(-D_MT=1)`

**Comment by rohit-n** [comment]
That fixed the issue. Shall I make a PR?

**Comment by DanielGibson** [comment]
I'm currently writing code to get rid of the dependency on UINTPTR_MAX, so maybe not ;)

**Comment by DanielGibson** [comment]
now it's like https://github.com/DanielGibson/Serious-Engine/commit/c1b14f9130303d7aac2189f2df8373cdc4108774
I'll not create a pull request yet, wanna do some more real fixes first

**Pull request merged**